### DISCUSSION TITLE
Next.js + Apollo Server 3.0 + Subscription

### DISCUSSION THREAD

**Discussion by nathantqn** [vercel/next.js#27680] [Help]
Hi guys, I successfully integrated Apollo Server 3.0 into Next.js route called /api by using below code snippet

```javascript
import { ApolloServer } from "apollo-server-micro"; // NOSONAR
import { typeDefs } from "../../graphql/schema";
import { resolvers } from "../../graphql/resolvers";

import Cors from "micro-cors";
import { API_URL } from "../../constants/common";

const cors = Cors();

const apolloServer = new ApolloServer({
  resolvers,
  typeDefs,
  introspection: true,
});

// @ts-ignore
await apolloServer.start();

export default cors(async (req, res) => {
  if (req.method === "OPTIONS") {
    res.end();
    return false;
  }

  return apolloServer.createHandler({ path: API_URL })(req, res);
});

export const config = {
  api: {
    bodyParser: false,
  },
};

```

Right now I want to integrate the Subscription functionality into this api, but I cannot get it working. 

This stackoverflow [topic](https://stackoverflow.com/questions/62371284/apollo-server-with-subscriptions-used-inside-next-js-api-routes-websockets-trou) is outdated, and the apollo-server-micro also removed the integrated subscription in the Apollo Server instance. 

If I tried to follow the example from Apollo [here](https://www.apollographql.com/docs/apollo-server/data/subscriptions/), The last step requires me to run a httpServer by following code:

```javascript
(async function () {
  const app = express();

  const httpServer = createServer(app);

  const schema = makeExecutableSchema({
    typeDefs,
    resolvers,
  });

  const server = new ApolloServer({
    schema,
  });
  await server.start();
  server.applyMiddleware({ app });

  SubscriptionServer.create(
    { schema, execute, subscribe },
    { server: httpServer, path: server.graphqlPath }
  );

  const PORT = 4000;
  httpServer.listen(PORT, () =>
    console.log(`Server is now running on http://localhost:${PORT}/graphql`)
  );
})();
```

Looking at the code snippet above, I have no idea how to convert it into Next.js's requestHandler.

Do you guys have any idea on this? Many thanks!

**Comment by tettoffensive** [comment]
I don't know the answer to your question, but [I've learned that over at Apollo](https://github.com/apollographql/apollo-server/issues/5547#issuecomment-929686003) they don't think we should be using `apollo-server-micro`.

That example needs to be updated.

(BTW. Would love to get subscriptions working myself too)

**Comment by meek-bracster** [comment]
@nenjamin2405  Did you ever figure out a solution for this?

**Comment by Mathieuka** [comment]
Based on the Apollo documentation, since `Apollo-V3` subscription is now handled by a different server , so `apollo-server-micro` no longer natively supports subscription 

[Apollo doc ](https://www.apollographql.com/docs/apollo-server/data/subscriptions/#enabling-subscriptions) : 
Beginning in Apollo Server 3, subscriptions are not supported by the "batteries-included" apollo-server package. To enable subscriptions, you must first swap to the apollo-server-express package (or any other Apollo Server integration package that supports subscriptions).

The following steps assume you've already swapped to apollo-server-express.

**Reply by tettoffensive** [reply]
FYI, https://github.com/vercel/next.js/pull/30082

**Reply by LModipane** [reply]
Is it possible to have an Apollo server using the express integration in the API folder for a next.js project?

**Comment by alimek** [comment]
anyone have this working or have solution for this use case ? using apollo server express in next js v12 atm without subscription and trying to add subscriptions into it

**Reply by tettoffensive** [reply]
Haven't gotten around to it. Are you having issues?

**Reply by alimek** [reply]
```ts
import { ApolloServer } from "apollo-server-express";
import schema from "../../graphql";
import { NextApiRequest, NextApiResponse } from "next";

const apolloServer = new ApolloServer({
  schema,
});

const runMiddleware = (req: NextApiRequest, res: NextApiResponse, fn) =>
  new Promise((resolve, reject) => {
    fn(req, res, (result) => {
      if (result instanceof Error) {
        return reject(result);
      }

      return resolve(result);
    });
  });

await apolloServer.start();
const apolloMiddleware = apolloServer.getMiddleware({
  path: "/api/graphql",
});

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse,
) {
  await runMiddleware(req, res, apolloMiddleware);
}

export const config = {
  api: {
    bodyParser: false,
  },
};
```

I am at the stage when `pollInternal` is OK for development in Apollo Client, but I would like to have subscription working or even [Live Query](https://www.the-guild.dev/blog/subscriptions-and-live-queries-real-time-with-graphql) later as I am working on multiplayer online game where I would like to have as up to date as possible data exchanged via socket.

Spent for now a day trying to understand how to integrate it with [graphql-ws](https://github.com/enisdenjo/graphql-ws), but didnt find solution.

**Reply by erezkastel** [reply]
@alimek did you ever find a solution for enabling subscriptions using apollo server express in next js ?

**Reply by alimek** [reply]
not really, I stopped doing a side project, so I didn't try it later, but in meantime I remember i was using long polling for that.

**Comment by Emad-Askari-Motlagh** [comment]
Here is my way with help of web socket and GraphQL-yoga.
Here is server:
`// import { createServer } from "@graphql-yoga/node";
import { makeExecutableSchema } from "@graphql-tools/schema";
import typeDefs from "@/server/graphql/typeDef/schema.graphql";
import resolvers from "@/server/graphql/resolvers";
import dbInit from "@/lib/dbInit";
import JWT from "jsonwebtoken";
import Cors from "micro-cors";

// const pubsub = new PubSub();
const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

import {
  createServer,
  createPubSub,
  GraphQLYogaError,
} from "@graphql-yoga/node";
import { useResponseCache } from "@envelop/response-cache";
import { WebSocketServer } from "ws"; // yarn add ws
// import ws from 'ws'; yarn add ws@7
// const WebSocketServer = ws.Server;
import { useServer } from "graphql-ws/lib/use/ws";

const pubSub = createPubSub();

const server = createServer({
  cors: {
    credentials: "same-origin",
    origin: ["http://localhost:3000"], // your frontend url.
  },

  plugins: [
    useResponseCache({
      includeExtensionMetadata: true,
    }),
  ],
  context: async (ctx) => {
    let wsServer = null;
    wsServer = ctx.res.socket.server.ws ||= new WebSocketServer({
      port: 4000,
      path: "/api/graphql",
    });
    wsServer &&= useServer({ schema }, wsServer);

    const db = await dbInit();
    let { token, costumerId, costumerExpire } = ctx.req.cookies;
    // 1. Find optional visitor id
    let id = null;
    if (token) {
      try {
        let obj = JWT.verify(token, "MY_SECRET");
        id = obj.id;
      } catch (err) {
        console.error("error on apollo server", err); // expired token, invalid token
        // TODO try apollo-link-error on the client
        throw new AuthenticationError(
          "Authentication token is invalid, please log in"
        );
      }
    }

    return {
      ...ctx,
      userId: id,
      costumerId,
      pubSub,
    };
  },
  schema,

});

export default server;
`
And client:
`import { useMemo } from "react";
import {
  ApolloClient,
  InMemoryCache,
  split,
  HttpLink,
  createHttpLink,
} from "@apollo/client";
import merge from "deepmerge";
import { getMainDefinition } from "apollo-utilities";
import { GraphQLWsLink } from "@apollo/client/link/subscriptions";
import { createClient } from "graphql-ws";

let apolloClient;
//create websocket link
const wsLink =
  typeof window !== "undefined"
    ? new GraphQLWsLink(
        createClient({
          url: "ws://localhost:4000/api/graphql",

          on: {
            connected: () => console.log("connected client"),
            closed: () => console.log("closed"),
          },
        })
      )
    : null;

//create http link
const httplink = new HttpLink({
  uri: "http://localhost:3000/api/graphql",
  credentials: "same-origin",
});

//Split the link based on graphql operation
const link =
  typeof window !== "undefined"
    ? split(
        //only create the split in the browser
        // split based on operation type
        ({ query }) => {
          const { kind, operation } = getMainDefinition(query);
          return kind === "OperationDefinition" && operation === "subscription";
        },
        wsLink,
        httplink
      )
    : httplink;

//create apollo client
function createApolloClient() {
  return new ApolloClient({
    ssrMode: typeof window === "undefined",
    link: link,
    cache: new InMemoryCache(),
  });
}

export function initializeApollo(initialState = null) {
  const _apolloClient = apolloClient ?? createApolloClient();

  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = merge(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }

  if (typeof window === "undefined") return _apolloClient;

  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState) {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
`

**Comment by stomvi** [comment]
I 'm using `next@12`, `apollo-server-micro@v3`, and `graphql-ws`. Here's my code:

```javascript
import { models } from '@elleshop/models';
import { ApolloServer } from 'apollo-server-micro';
import { useServer } from 'graphql-ws/lib/use/ws';
import { WebSocketServer } from 'ws';

import schema from '../../server/schema';

export const config = {
  api: {
    bodyParser: false,
  },
};

const isDev = process.env.NODE_ENV !== consts.ENV_PRODUCTION;

// Apollo Server --------------------------------------------------------------
const apolloServer = new ApolloServer({
  schema,
  debug: isDev,
  cache: 'bounded',
  context({ req }) {
    const { session, socket } = req;
    return { models, session, pubsub: socket?.server?.pubsub };
  },
  plugins: [
    {
      async serverWillStart(...asdf) {
        return {
          async drainServer() {
            console.log(`ApolloServer: drain server :>>>>>>>>>>>>>>>>>>`);
            if (graphqlWSS) {
              console.log(`ApolloServer: disposing WSS :>>>>>>>>>>>>>>>>>>`);
              await graphqlWSS.dispose();
            }
          },
        };
      },
    },
  ],
});

const startServer = apolloServer.start();
let graphqlWSS;
let apolloServerHandler;

const handler = async (req, res) => {
  if (req.method === 'OPTIONS') {
    return res.end();
  }

  const oldApolloServer = res.socket.server.apolloServer;

  if (oldApolloServer && oldApolloServer !== apolloServer) {
    console.warn('Fixing Apollo Server hot reload');
    oldApolloServer.stop();
    delete res.socket.server.apolloServer;
  }

  if (!res.socket.server.apolloServer) {
    res.socket.server.apolloServer = apolloServer;
    if (!graphqlWSS) {
      /* eslint-disable react-hooks/rules-of-hooks */
      console.log(`Initializing GraphQL WSS :>>>>>>>>>>>>>>>>>>`);
      const wss = new WebSocketServer({
        server: res.socket.server,
        path: '/api/graphql',
      });

      graphqlWSS = useServer(
        {
          schema,
          context() {
            const { session, socket } = req;
            return { session, pubsub: socket.server.pubsub };
          },
        },
        wss,
      );
    }

    await startServer;

    apolloServerHandler = apolloServer.createHandler({ path: '/api/graphql' });
  }

  await apolloServerHandler(req, res);
};

export default handler;
```

The cleanup plugin is inspired by [https://stackoverflow.com/a/66268643](https://stackoverflow.com/a/66268643).

This code is working but not fully tested. Use with caution.

**Reply by stomvi** [reply]
You will have to setup client side network as well: https://relay.dev/docs/guided-tour/updating-data/graphql-subscriptions/

**Reply by wonecode** [reply]
Hi stomvi, did you find out a stable version with your code ? I tried to use it but it looks there is some problems with cors (like 'access-control-allow-origin'). Could you maybe please help me ?

**Reply by stomvi** [reply]
> Hi stomvi, did you find out a stable version with your code ? I tried to use it but it looks there is some problems with cors (like 'access-control-allow-origin'). Could you maybe please help me ?

What problem and where does it occur?

**Reply by wonecode** [reply]
This is my code for the server, inside `/pages/api/graphql.ts` , but the apollo studio cannot reach the server, it can only reach the subscription server.

```js
import { ApolloServer } from 'apollo-server-micro';
import { useServer } from 'graphql-ws/lib/use/ws';
import { WebSocketServer } from 'ws';
import resolvers from '@/graphql/resolvers/resolvers';
import typeDefs from '@/graphql/schemas/typeDefs';
import { prisma } from '@/server/prismaClient';
import { makeExecutableSchema } from '@graphql-tools/schema';

const schema = makeExecutableSchema({ typeDefs, resolvers });

export const config = {
  api: {
    bodyParser: false,
    externalResolver: true,
  },
};

// Apollo Server --------------------------------------------------------------
const apolloServer = new ApolloServer({
  schema,
  cache: 'bounded',
  context({ req }) {
    const { socket } = req;
    return { pubsub: socket?.server?.pubsub, prisma };
  },
  plugins: [
    {
      async serverWillStart() {
        return {
          async drainServer() {
            console.log(`ApolloServer: drain server :>>>>>>>>>>>>>>>>>>`);
            if (graphqlWSS) {
              console.log(`ApolloServer: disposing WSS :>>>>>>>>>>>>>>>>>>`);
              await graphqlWSS.dispose();
            }
          },
        };
      },
    },
  ],
});

const startServer = apolloServer.start();
let graphqlWSS;
let apolloServerHandler;

const handler = async (req: any, res: any) => {
  if (req.method === 'OPTIONS') {
    return res.end();
  }

  if (req.headers) {
    console.log(`ApolloServer: headers :>>>>>>>>>>>>>>>>>>`);
    console.log(req.headers);
  }

  const oldApolloServer = res.socket.server.apolloServer;

  if (oldApolloServer && oldApolloServer !== apolloServer) {
    console.warn('Fixing Apollo Server hot reload');
    oldApolloServer.stop();
    delete res.socket.server.apolloServer;
  }

  if (!res.socket.server.apolloServer) {
    res.socket.server.apolloServer = apolloServer;

    if (!graphqlWSS) {
      /* eslint-disable react-hooks/rules-of-hooks */
      console.log(`Initializing GraphQL WSS :>>>>>>>>>>>>>>>>>>`);
      const wss = new WebSocketServer({
        server: res.socket.server,
        path: '/api/graphql',
      });

      graphqlWSS = useServer(
        {
          schema,
          context() {
            const { socket } = req;
            return { pubsub: socket.server.pubsub };
          },
        },
        wss
      );
    }

    await startServer;

    apolloServerHandler = apolloServer.createHandler({ path: '/api/graphql' });
  }

  await apolloServerHandler(req, res);
};

export default handler;
```

I tried a `npx diagnose-endpoint@1.1.0 --endpoint=http://localhost:3000/api/graphql` and it returns : 

```bash
Diagnosing http://localhost:3000/api/graphql
⚠️  OPTIONS response is missing header 'access-control-allow-methods: POST'
⚠️  POST response missing 'access-control-allow-origin' header.
If using cookie-based authentication, the following headers are required from your endpoint: 
    access-control-allow-origin: https://studio.apollographql.com
    access-control-allow-credentials: true
Otherwise, a wildcard value would work:
    access-control-allow-origin: *
(📫 Interested in previewing a local tunnel to bypass CORS requirements? Please let us know at https://docs.google.com/forms/d/e/1FAIpQLScUCi3PdMerraiy6GpD-QiC_9KEKVHr4oDL5Vef5fIvzqqQWg/viewform )
```

This is also my code for the client, inside `_app.tsx` : 
```jsx
import { ApolloClient, ApolloProvider, InMemoryCache } from '@apollo/client';
import { CssBaseline } from '@mui/material';
import { ThemeProvider } from '@mui/material/styles';
import { SessionProvider } from 'next-auth/react';
import Router from 'next/router';
import NProgress from 'nprogress';
import { Provider } from 'react-redux';
import { ToastContainer } from 'react-toastify';

import { HttpLink, split } from '@apollo/client';
import { GraphQLWsLink } from '@apollo/client/link/subscriptions';
import { getMainDefinition } from '@apollo/client/utilities';
import { createClient } from 'graphql-ws';

import store from '../redux/store';
import { hmtTheme } from '../styles/theme';

// Styles
import 'nprogress/nprogress.css';
import 'react-toastify/dist/ReactToastify.css';
import 'tippy.js/dist/tippy.css';
import '../styles/globals.css';

const httpLink = new HttpLink({
  uri: '/api/graphql',
});

const wsLink =
  typeof window !== 'undefined'
    ? new GraphQLWsLink(
        createClient({
          url: 'ws://localhost:3000/api/graphql',
        })
      )
    : null;

const link =
  typeof window !== 'undefined' && wsLink != null
    ? split(
        ({ query }) => {
          const def = getMainDefinition(query);
          return (
            def.kind === 'OperationDefinition' &&
            def.operation === 'subscription'
          );
        },
        wsLink,
        httpLink
      )
    : httpLink;

const client = new ApolloClient({
  cache: new InMemoryCache({
    typePolicies: {
      ChatConversation: {
        fields: {
          messages: {
            merge(
              existing: any[],
              incoming: any[],
              { readField, mergeObjects }
            ) {
              const merged: any[] = existing ? existing.slice(0) : [];
              const messageToIndex: Record<string, number> =
                Object.create(null);
              if (existing) {
                existing.forEach((message, index) => {
                  messageToIndex[readField<string>('id', message)] = index;
                });
              }
              incoming.forEach((message) => {
                const id = readField<string>('id', message);
                const index = messageToIndex[id];
                if (typeof index === 'number') {
                  // Merge the new author data with the existing author data.
                  merged[index] = mergeObjects(merged[index], message);
                } else {
                  // First time we've seen this author in this array.
                  messageToIndex[id] = merged.length;
                  merged.push(message);
                }
              });
              return merged;
            },
          },
        },
      },
    },
  }),
  link,
  uri: '/api/graphql',
});

NProgress.configure({ showSpinner: false });

Router.events.on('routeChangeStart', () => NProgress.start());
Router.events.on('routeChangeComplete', () => NProgress.done());
Router.events.on('routeChangeError', () => NProgress.done());

function App({ Component, pageProps: { session, ...pageProps } }) {
  return (
    <ApolloProvider client={client}>
      <SessionProvider session={session}>
        <Provider store={store}>
          <ThemeProvider theme={hmtTheme}>
            <ToastContainer
              containerId="app"
              draggable={false}
              position="bottom-right"
              toastStyle={{ backgroundColor: '#252525', color: '#ecf0f1' }}
            />
            <CssBaseline />
            <Component {...pageProps} />
          </ThemeProvider>
        </Provider>
      </SessionProvider>
    </ApolloProvider>
  );
}

export default App;
```

I've been on it for 3 days and I'm really breaking my head... I hope you can help me find the solution, thanks a lot in advance

**Reply by stomvi** [reply]
This might help:
[https://www.apollographql.com/docs/studio/explorer/connecting-authenticating/](https://www.apollographql.com/docs/studio/explorer/connecting-authenticating/)

**Reply by stomvi** [reply]
And this:
[https://expressjs.com/en/resources/middleware/cors.html](https://expressjs.com/en/resources/middleware/cors.html)

If you are using a Custom Server with `express`, you could use `cors` package to do the trick. If you aren't, check this out:
[https://www.npmjs.com/package/nextjs-cors](https://www.npmjs.com/package/nextjs-cors)

**Reply by stomvi** [reply]
For those create a WebSocket server in Next.js API Route and lose HMR, this might help:
[https://stackoverflow.com/questions/69840693/how-can-i-run-a-websocket-server-in-next-js-custom-server-in-dev-mode](https://stackoverflow.com/questions/69840693/how-can-i-run-a-websocket-server-in-next-js-custom-server-in-dev-mode)

**Reply by wonecode** [reply]
Thanks a lot, i'll get a look !

**Reply by stomvi** [reply]
Currently I move out all my `graphql-ws` code into the Custom Server file since ws server has to respond to `http` server event to keep the HMR working.  I wonder if there is any more elegant way to do so.

**Reply by r7d9n4** [reply]
can you share please your custom server if it working, cause with this solution i'm stuck with this error WebSocket connection to 'ws://localhost:3000/_next/webpack-hmr' failed: Invalid frame header

**Comment by theEngMansour** [comment]
How added subscription with graphql to next.js

**Comment by stomvi** [comment]
Hasn't updated for a while. I just jumped into `graphql-yoga`. I found their `custom server` example shared the same concept as mine. Check it out:

[https://the-guild.dev/graphql/yoga-server/docs/integrations/integration-with-nextjs](https://the-guild.dev/graphql/yoga-server/docs/integrations/integration-with-nextjs)
